client/asset/eth/eth.go

```diff
@@ -1530,73 +2482,177 @@ func (eth *ExchangeWallet) balanceWithTxPool() (*Balance, error) {
-	return gas, eth.withContractor(contractVer, func(c contractor) error {
+func (w *AssetWallet) estimateRefundGas(ctx context.Context, secretHash [32]byte, contractVer uint32) (gas uint64, err error) {
+	return gas, w.withContractor(contractVer, func(c contractor) error {
 		gas, err = c.estimateRefundGas(ctx, secretHash)
 		return err
 	})
 }
 
+// loadContractors prepares the token contractors and add them to the map.
+func (w *AssetWallet) loadContractors(ctx context.Context, tokenID uint32) error {
+	token, found := dexeth.Tokens[tokenID]
+	if !found {
+		return fmt.Errorf("token %d not found", tokenID)
+	}
+	netToken, found := token.NetTokens[w.net]
+	if !found {
+		return fmt.Errorf("token %d not found", tokenID)
+	}
+
+	for ver := range netToken.SwapContracts {
+		constructor, found := tokenContractorConstructors[ver]
+		if !found {
+			w.log.Errorf("contractor constructor not found for token %s, version %d", token.Name, ver)
+			continue
+		}
+		c, err := constructor(w.net, tokenID, w.addr, w.node.contractBackend())
+		if err != nil {
+			return fmt.Errorf("error constructing token %s contractor version %d: %w", token.Name, ver, err)
+		}
+
+		if netToken.Address != c.tokenAddress() {
+			return fmt.Errorf("wrong %s token address. expected %s, got %s", token.Name, netToken.Address, c.tokenAddress())
+		}
+
+		w.contractors[ver] = c
+	}
+	return nil
+}
+
+// withContractor runs the provided function with the versioned contractor.
+func (w *AssetWallet) withContractor(contractVer uint32, f func(contractor) error) error {
+	if contractVer == contractVersionNewest {
+		var bestVer uint32
+		var bestContractor contractor
+		for ver, c := range w.contractors {
+			if ver >= bestVer {
+				bestContractor = c
+				bestVer = ver
+			}
+		}
+		return f(bestContractor)
+	}
+	contractor, found := w.contractors[contractVer]
+	if !found {
+		return fmt.Errorf("no version %d contractor for asset %d", contractVer, w.assetID)
+	}
+	return f(contractor)
+}
+
+// withTokenContractor runs the provided function with the tokenContractor.
+func (w *AssetWallet) withTokenContractor(assetID, ver uint32, f func(tokenContractor) error) error {
+	return w.withContractor(ver, func(c contractor) error {
+		tc, is := c.(tokenContractor)
+		if !is {
+			return fmt.Errorf("contractor for %d %T is not a tokenContractor", assetID, c)
+		}
+		return f(tc)
+	})
+}
+
+// estimateApproveGas estimates the gas required for a transaction approving a
+// spender for an ERC20 contract.
+func (w *AssetWallet) estimateApproveGas(newGas *big.Int) (gas uint64, err error) {
+	return gas, w.withTokenContractor(w.assetID, contractVersionNewest, func(c tokenContractor) error {
+		gas, err = c.estimateApproveGas(w.ctx, newGas)
+		return err
+	})
+}
+
+// estimateTransferGas estimates the gas needed for a token transfer call to an
+// ERC20 contract.
+func (w *AssetWallet) estimateTransferGas(val uint64) (gas uint64, err error) {
```

Comment:
These three look like they should be off of `TokenWallet`.

Reply:
Damn. I can't do this because of some other shared code. Maybe it would be worth splitting more methods up, but I'm not feeling it. Can follow up.